oeps/oep-38-Data-Modeling.rst

```diff
@@ -0,0 +1,438 @@
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Data stored in the OpenEdX ecosystem should adhere to industry best
+practices. For example, since an industry best practice uses numerical
+identifiers to identify rows of data, the edX ecosystem should also use
+numerical identifiers. Adhering to industry best practices and the
+practices outlined in this document ensures our data is approachable for
+experienced engineers and new hires alike and allows us to leverage 3rd
+party tools more easily to assist in the analysis phase.
+
+Don’t make me think
+~~~~~~~~~~~~~~~~~~~
+
+Steve Krug, the author of “Don’t Make Me Think”, says: “Your objective
+should always be to eliminate instructions entirely by making everything
+self-explanatory, or as close to it as possible.” This principal should
+also be applied to data modeling. Data models should have descriptive
+names clearly identifying the data that the model holds. The field names
+should make sense to most people familiar with the domain, without
+needing much clarifying documentation.
+
+Prioritize ease of analysis
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+For every hour of engineer-time spent creating a data model, several
+thousand more are spent using the model and analyzing its data. Taking
+the time to consider how the data will be used, and thoughtfully design
+our data representation, can potentially save many hours during the
+analysis phase.
+
+This is especially true since changing data models, once they are in
+use, can be time-consuming. A data model change in a core application
+can affect many other systems. Small changes to a data model could
+possibly cause hours or days of work for different teams throughout the
+organization:
+
+-  Data engineering and Analytics to update their workflows
+
+-  BI teams to update reports that utilize this data,
+
+-  Devops to deploy migrations to large tables, with the risk of
+      potential downtime.
+
+Due to the potentially high cost of changing a model, it pays to get it
+as correct as possible the first time.
+
+Think about the person who will analyze this data later and the person
+who manages the system day to day. (That person could possibly be future
+you!) Try to do as much careful design up front to make your quality of
+life better later.
+
+History is best (but events are a close second)
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Accuracy is best when history about a change is captured at the database
+level when the change happens. Changes in data relating to business
+continuity should be stored historically where the change is made,
+preferably in the same system as the system of record. This allows us to
+reconstruct the data at different points with much greater reliability.
+
+If, for some reason, it is not possible to create history at the time of
+entry (for example, if we expect the table to become too large or the
+writes to be too frequent), We recommend that an event be issued in its
+place.
+
+Fire events for CUD operations
```

Comment:
Are there any recommendations for eventing best-practices at edX?

Reply:
This is the only doc I know of: https://openedx.atlassian.net/wiki/spaces/AN/pages/13205895/Event+Design+and+Review+Process

Reply:
Review decision: We will do a separate OEP for eventing.